src/hotspot/share/utilities/checkedCast.hpp

```diff
@@ -25,22 +25,177 @@
-  return result;
+// Implementation support for checked_cast.
+//
+// Because of details of C++ conversion rules, including undefined or
+// implementation-defined behavior in some cases, along with the possibility
+// of tautological comparison warnings from some compilers, the implementation
+// is not as simple as doing the obvious range checks.
+class CheckedCastImpl {
+  template<typename To, typename From, bool permit_tautology>
+  struct CheckSameSign;
+
+  template<typename To, typename From, bool permit_tautology,
+           bool is_narrowing = (sizeof(To) <= sizeof(From))>
+  struct CheckToSigned;
+
+  template<typename To, typename From, bool permit_tautology,
+           bool is_narrowing = (sizeof(To) < sizeof(From))>
+  struct CheckFromSigned;
+
+  template<typename To, typename From>
+  static constexpr bool check_range(From from) {
+    To to_max = std::numeric_limits<To>::max();
+    return from <= static_cast<From>(to_max);
+  }
+
+  template<typename To, typename From, bool permit_tautology>
+  struct Dispatcher {
+    static const bool to_signed = std::is_signed<To>::value;
+    static const bool from_signed = std::is_signed<From>::value;
+    using SameSign = CheckSameSign<To, From, permit_tautology>;
+    using ToSigned = CheckToSigned<To, From, permit_tautology>;
+    using FromSigned = CheckFromSigned<To, From, permit_tautology>;
+    using type =
+      std::conditional_t<(to_signed == from_signed),
+                         SameSign,
+                         std::conditional_t<to_signed, ToSigned, FromSigned>>;
+  };
+
+public:
+  template<typename To, typename From>
+  static constexpr bool is_tautology() {
+    using Dispatcher = Dispatcher<To, From, true>;
+    using Checker = typename Dispatcher::type;
+    return Checker::is_tautology;
+  }
+
+  template<typename To, bool permit_tautology, typename From>
+  static constexpr bool check(From from) {
+    using Dispatcher = Dispatcher<To, From, permit_tautology>;
+    using Checker = typename Dispatcher::type;
+    return Checker()(from);
+  }
+};
+
+// If both types are signed or both are unsigned, only a narrowing conversion
+// can lose information.  We check for such loss via a round-trip conversion.
+// C++14 4.7 defines the result for unsigned types, and implementation-defined
+// for signed types.  All supported implementations do the "obvious" discard
+// of high-order bits when narrowing signed values, and C++20 requires that
+// behavior.  This approach avoids any additional complexity to avoid
+// potential tautological comparisons.
+template<typename To, typename From, bool permit_tautology>
+struct CheckedCastImpl::CheckSameSign {
+  static const bool is_narrowing = sizeof(To) < sizeof(From);
+  static const bool is_tautology = !is_narrowing;
+  static_assert(permit_tautology || is_narrowing, "tautological checked_cast");
+  constexpr bool operator()(From from) const {
+    return !is_narrowing || (static_cast<From>(static_cast<To>(from)) == from);
```

Comment:
OK, so we are allowing compiler-specific behavior now?  Does that mean we can get rid of the weird-looking `reinterpret_cast<TYPE&>(ures)` in `JAVA_INTEGER_OP`?